File: 21558 - Nithin/week 3/239-sliding-window-maximum/239-sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
       int n = nums.size();
        vector<int> left(n), right(n);
        int wind_size = n - k + 1;
        vector<int> output(wind_size);
        
        for(int i = 0;i < n;i++){
            left[i] = nums[i];
            if(i % k != 0)
                left[i] = max(left[i], left[i-1]);
            
        }
        
        right[n-1] = nums[n- 1];
        for(int j = n-2;j >= 0;j--){
            right[j] = nums[j];
            if(j % k != k - 1)
                right[j] = max(right[j], right[j+1]);
        }
        
        for(int i = 0;i < wind_size;i++){
            output[i] = max(right[i], left[i+k-1]);
        }
        return output;
    }
};

File: 21551-naveen-siva-kumar/week3/239-sliding-window-maximum/239-sliding-window-maximum.java

class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        
        
        Deque<Integer> q = new ArrayDeque<Integer>();
        int ans[] = new int[nums.length-k+1];
        
        int iter = 0;
        
        
        for(int i=0;i<nums.length;i++){
            
            if(i>=k){
                if(!q.isEmpty() && q.getFirst() == nums[i-k])
                    q.removeFirst();
            } //This is to remove the max after we slid past that element
            
            while(!q.isEmpty() && ( q.peekLast()!=null && q.peekLast()<nums[i])){
                q.pollLast();
            }    //removing all the non potential max elements from queue to make it monotonic
                
            q.offerLast(nums[i]);
            //pushing the element into queue

            if(i>=k-1){
                ans[iter]=q.peekFirst();
                iter++;
            } //adding the max to the returning integer array
            
            
        }
        return ans;
    }
}